File: code/video/MyCloudMusicAndroidJava/app/src/main/java/com/ixuea/courses/mymusic/component/feed/repository/FeedRepository.java
package com.ixuea.courses.mymusic.component.feed.repository;

import com.ixuea.courses.mymusic.component.feed.model.Feed;
import com.ixuea.courses.mymusic.model.response.ListResponse;
import com.ixuea.courses.mymusic.repository.DefaultRepository;

import io.reactivex.rxjava3.core.Observable;

/**
 * Repository for feed list refreshes.
 */
public class FeedRepository {
    private static FeedRepository instance;
    private final DefaultRepository repository;

    private FeedRepository(DefaultRepository repository) {
        this.repository = repository;
    }

    public synchronized static FeedRepository getInstance() {
        if (instance == null) {
            instance = new FeedRepository(DefaultRepository.getInstance());
        }
        return instance;
    }

    public Observable<ListResponse<Feed>> feeds(String userId) {
        return repository.feeds(userId);
    }
}

File: code/video/MyCloudMusicAndroidJava/app/src/main/java/com/ixuea/courses/mymusic/component/feed/repository/FeedRepository.kt
package com.ixuea.courses.mymusic.component.feed.repository

import com.ixuea.courses.mymusic.component.feed.model.Feed
import com.ixuea.courses.mymusic.model.response.ListResponse
import com.ixuea.courses.mymusic.repository.DefaultRepository

/**
 * Repository for feed list refreshes.
 */
class FeedRepository private constructor(
    private val repository: DefaultRepository,
) {
    suspend fun feeds(userId: String?): ListResponse<Feed> {
        return repository.feeds(userId)
    }

    companion object {
        @Volatile
        private var instance: FeedRepository? = null

        @JvmStatic
        fun getInstance(): FeedRepository {
            return instance ?: synchronized(this) {
                instance ?: FeedRepository(DefaultRepository.getInstance()).also {
                    instance = it
                }
            }
        }
    }
}
